onmt/modules/Transformer.py

```diff
@@ -55,60 +35,45 @@ def __init__(self, hidden_size, dropout, padding_idx,
         self.feed_forward = PositionwiseFeedForward(hidden_size,
                                                     d_inner,
                                                     dropout)
-        self.padding_idx = padding_idx
-
-    def forward(self, input, words):
-        # CHECKS
-        n_batch, s_len, _ = input.size()
-        n_batch_, s_len_ = words.size()
-        aeq(n_batch, n_batch_)
-        aeq(s_len, s_len_)
-        # END CHECKS
 
-        mask = get_attn_padding_mask(words, words, self.padding_idx)
+    def forward(self, input, mask):
         mid, _ = self.self_attn(input, input, input, mask=mask)
         out = self.feed_forward(mid)
         return out
 
 
 class TransformerDecoder(nn.Module):
     """
-    The Transformer Decoder from AIAYN
+    The Transformer Decoder from paper "Attetion is all you need".
     """
-    def __init__(self, hidden_size, opt, pad,
+    def __init__(self, hidden_size, dropout,
                  n_head=8, d_inner=2048):
         super(TransformerDecoder, self).__init__()
         self.self_attn = onmt.modules.MultiHeadedAttention(n_head, hidden_size,
-                                                           p=opt.dropout)
+                                                           p=dropout)
         self.context_attn = onmt.modules.MultiHeadedAttention(n_head,
                                                               hidden_size,
-                                                              p=opt.dropout)
+                                                              p=dropout)
         self.feed_forward = PositionwiseFeedForward(hidden_size,
                                                     d_inner,
-                                                    opt.dropout)
-        self.dropout = opt.dropout
-        self.mask = get_attn_subsequent_mask(5000)
-        if len(opt.gpuid) > 0:
-            self.mask = self.mask.cuda()
-        self.pad = pad
-
-    def forward(self, input, context, src_words, tgt_words):
+                                                    dropout)
+        self.dropout = dropout
+        self.mask = self._get_attn_subsequent_mask(5000)
+        # Register self.mask as a parameter of TransformerDecoder, so
+        # it gets TransformerDecoder's cuda behavior automatically.
+        self.register_parameter('mask',
```

Comment:
I think this should be a `buffer` not a `parameter`. See what was done in the new PR for self.pe in embeddings.

Reply:
Okay.